### PR TITLE
[16.0][ADD] mrp_unbuild_source_location: filter source location by lot/serial stock

### PR DESCRIPTION
**Pull request by nobuQuartile**
When a lot/serial number is selected on an unbuild order, restrict the available source locations to those where the lot has stock. Auto-fill the location when only one option exists, and show a warning if no stock is found.

@qrtl
QT6115

### CODE REVIEW

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                            ("product_id", "=", record.product_id.id),
+                            ("lot_id", "=", record.lot_id.id),
+                            ("quantity", ">", 0),
+                            ("location_id.usage", "in", ["internal", "transit"]),
+                        ]
+                    )
+                )
+                record.location_id_domain = quants.mapped("location_id")
+            else:
+                record.location_id_domain = self.env["stock.location"]
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]:
```suggestion
```

File: mrp_unbuild_lot_location/readme/CONTRIBUTORS.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+* `Quartile <https://www.quartile.co>`_:
+
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/readme/CONTRIBUTORS.rst]:
```suggestion
```

File: mrp_unbuild_lot_location/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2025 Quartile
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "MRP Unbuild Lot Location",
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/__manifest__.py:4]: Please add summary.

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from odoo import _, api, fields, models
+
+
+class MrpUnbuild(models.Model):
+    _inherit = "mrp.unbuild"
+
+    location_id_domain = fields.Many2many(
+        "stock.location",
+        compute="_compute_location_id_domain",
+    )
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]: How about using Binary field?

```suggestion
    location_id_domain = fields.Binary(compute="_compute_location_id_domain")
```

File: mrp_unbuild_lot_location/models/mrp_unbuild.py (hunk 1 above)
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]:
```suggestion
    @api.depends("lot_id", "product_id", "company_id")
    def _compute_location_id_domain(self):
        for record in self:
            base_domain = [
                ("usage", "in", ["internal", "transit"]),
                "|",
                ("company_id", "=", False),
                ("company_id", "=", record.company_id.id),
            ]
            if record.lot_id:
                quants = self.env["stock.quant"].search(
                    [
                        ("product_id", "=", record.product_id.id),
                        ("lot_id", "=", record.lot_id.id),
                        ("quantity", ">", 0),
                        ("location_id.usage", "in", ["internal", "transit"]),
                    ]
                )
                base_domain = expression.AND(
                    [base_domain, [("id", "in", quants.mapped("location_id").ids)]]
                )
            record.location_id_domain = base_domain
```

File: mrp_unbuild_lot_location/views/mrp_unbuild_views.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    <record id="mrp_unbuild_form_view_inherit_location" model="ir.ui.view">
+        <field name="name">mrp.unbuild.form.inherit.location</field>
+        <field name="model">mrp.unbuild</field>
+        <field name="inherit_id" ref="mrp.mrp_unbuild_form_view" />
+        <field name="arch" type="xml">
+            <field name="location_id" position="before">
+                <field name="location_id_domain" invisible="1" />
+            </field>
+            <field name="location_id" position="attributes">
+                <attribute
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/views/mrp_unbuild_views.xml]: Shouldn't we put the existing conditions in compute method and just simply add below?
`<attribute name="domain">location_id_domain</attribute>`

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            return
+        if len(self.location_id_domain) == 1:
+            self.location_id = self.location_id_domain
+        elif not self.location_id_domain:
+            return {
+                "warning": {
+                    "message": _("No stock found for lot/serial number %s.")
+                    % self.lot_id.name,
+                }
+            }
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]: Shouldn't we just @api.constrains for this?

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    @api.depends("lot_id", "product_id", "company_id")
+    def _compute_location_id_domain(self):
+        for record in self:
+            base_domain = [
+                ("usage", "in", ["internal", "transit"]),
+                "|",
+                ("company_id", "=", False),
+                ("company_id", "=", record.company_id.id),
+            ]
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py:23]: Better to have a commend about this is the standard domains.

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2026 Quartile
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]: Please add company website link.

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+                    .search(
+                        [
+                            ("product_id", "=", record.product_id.id),
+                            ("lot_id", "=", record.lot_id.id),
+                            ("quantity", ">", 0),
+                        ]
+                    )
+                    .location_id.ids
+                )
+                base_domain += [("id", "in", location_ids)]
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]: It would be better to use `expression.AND`.

File: mrp_unbuild_lot_location/models/mrp_unbuild.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            record.location_id_domain = base_domain
+
+    @api.onchange("lot_id")
+    def _onchange_lot_id(self):
+        if not self.lot_id:
+            return
+        locations = self.env["stock.location"].search(self.location_id_domain)
+        if len(locations) == 1:
+            self.location_id = locations
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/models/mrp_unbuild.py]: I think this will get wrong result if `_compute_location_id_domain` is triggerd after this onchange method.
I am not sure odoo will trigger in which procedure. May be compute method first but current approach is based on the order. Can you please consider extending `_compute_location_id` by adding `location_id_domain` as a dependency?

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
+# Copyright 2026 Quartile
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo.exceptions import ValidationError
+from odoo.tests.common import TransactionCase
+
+
+class TestMrpUnbuildLotLocation(TransactionCase):
+    def setUp(self):
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: Use setupClass.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        super().setUp()
+        self.stock_location = self.env.ref("stock.warehouse0").lot_stock_id
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: Please avoid using demo data from odoo standard.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                "name": "Test Shelf 2",
+                "location_id": cls.warehouse.lot_stock_id.id,
+                "usage": "internal",
+            }
+        )
+        cls.product = cls.env["product.product"].create(
+            {"name": "Test Product", "type": "product", "tracking": "lot"}
+        )
+        cls.lot_single = cls.env["stock.lot"].create(
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: May be lot_1, lot_2 is better instead of single, multiple.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        """When lot_id is not set, domain returns all internal/transit locations."""
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertIn(self.test_location_1, locations)
+        self.assertIn(self.test_location_2, locations)
+
+    def test_onchange_lot_id_multiple_locations(self):
+        """When lot has stock in multiple locations, location_id is not auto-filled."""
+        self.unbuild.lot_id = self.lot_multiple
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertEqual(len(locations), 2)
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: Shouldn't we also check test_location_1 and test_location_2 are inside locations?

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        )
+        cls.unbuild = cls.env["mrp.unbuild"].create({"product_id": cls.product.id})
+
+    def test_location_id_domain_no_lot(self):
+        """When lot_id is not set, domain returns all internal/transit locations."""
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertIn(self.test_location_1, locations)
+        self.assertIn(self.test_location_2, locations)
+
+    def test_onchange_lot_id_multiple_locations(self):
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: Please remove onchange from method name since we don't use onchange method anymore.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        """When lot has stock in multiple locations, location_id is not auto-filled."""
+        self.unbuild.lot_id = self.lot_multiple
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertEqual(len(locations), 2)
+
+    def test_onchange_lot_id_single_location(self):
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: We should combine this test into above one.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        """When lot has stock in one location, location_id is auto-filled."""
+        self.unbuild.lot_id = self.lot_single
+        self.assertEqual(self.unbuild.location_id._origin, self.test_location_1)
+
+    def test_check_lot_location_no_stock(self):
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]: We should combine this test into above one.

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        # When lot_id is not set, domain returns all internal/transit locations.
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertIn(self.location_1, locations)
+        self.assertIn(self.location_2, locations)
+
+    def test_lot_location_resolution(self):
+        # When lot has stock in one location, location_id is auto-filled.
+        self.unbuild.lot_id = self.lot_1
+        locations = self.env["stock.location"].search(self.unbuild.location_id_domain)
+        self.assertEqual(self.unbuild.location_id._origin, self.location_1)
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]:
```suggestion
        self.assertEqual(self.unbuild.location_id, self.location_1)
```

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+                "name": "lot_3",
+                "product_id": cls.product.id,
+                "company_id": cls.env.company.id,
+            }
+        )
+        cls.env["stock.quant"]._update_available_quantity(
+            cls.product, cls.location_1, 5, lot_id=cls.lot_1
+        )
+        cls.env["stock.quant"]._update_available_quantity(
+            cls.product, cls.location_2, 3, lot_id=cls.lot_2
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]:
```suggestion
            cls.product, cls.location_1, 2, lot_id=cls.lot_2
```

File: mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        )
+        cls.env["stock.quant"]._update_available_quantity(
+            cls.product, cls.location_1, 2, lot_id=cls.lot_2
```
Comment (AungKoKoLin1997) [inline mrp_unbuild_lot_location/tests/test_mrp_unbuild_lot_location.py]:
```suggestion
            cls.product, cls.location_2, 3, lot_id=cls.lot_2
```

**Review by AungKoKoLin1997 — APPROVED** [review]
Code Review: LGTM
Minor points.

**Review by marcos-mendez — CHANGES_REQUESTED** [review]
**Automated Review -- Tests Failed**

### 1. **Root cause** of the test failure

The test failure occurs because the `_compute_location_id` method in `MrpUnbuild` overrides the parent's method but does not call `super()._compute_location_id()` properly. As a result, the `location_id` field is not correctly computed in all cases, especially when no lot is set or when stock is not found.

### 2. **Suggested fix**

In `mrp_unbuild_lot_location/models/mrp_unbuild.py`, modify the `_compute_location_id` method to ensure it correctly calls `super()` and handles the logic for setting `location_id`:

```python
@api.depends("company_id", "location_id_domain")
def _compute_location_id(self):
    super()._compute_location_id()  # Ensure parent logic runs first
    for record in self.filtered("lot_id"):
        locations = self.env["stock.location"].search(record.location_id_domain)
        if len(locations) == 1:
            record.location_id = locations
```

Also, fix the test case by ensuring `self.unbuild.lot_id = self.lot_3` is followed by a proper context refresh or recompute, since the `location_id` is not directly updated in the test but computed via domain.

### 3. **Additional code issues**

- **Missing `return` in `_compute_location_id`**: The method currently returns `None` implicitly, but it should be clear that it's not meant to return anything, or explicitly return `None` for clarity.
- **Incorrect test assertion**: In `test_lot_location_resolution`, the last assertion (`with self.assertRaises(ValidationError):`) is incomplete. It should include a call to save the record or trigger validation to actually test the constraint.

### 4. **Test improvements**

To improve test coverage, consider the following additions to `test_mrp_unbuild_lot_location.py`:

- Add a test case for when `lot_id` is set, but no stock exists (should raise `ValidationError`).
- Add a test case to ensure that `location_id` is not auto-filled when multiple locations have stock for the same lot.
- Add a test for `location_id_domain` with a company filter to ensure the domain respects company boundaries.

Example test case:
```python
def test_lot_no_stock_validation(self):
    # Test that ValidationError is raised when lot has no stock
    with self.assertRaises(ValidationError):
        self.unbuild.lot_id = self.lot_3
        self.unbuild._check_lot_location()
```

Use `TransactionCase` for full Odoo environment setup and `SavepointCase` if performance is a concern for more complex tests. Tag tests appropriately using `@tag('post_install', 'functional')` for tests that depend on module installation.

---
### Reciprocal Review Request



Hi everyone! I found some test failures on this PR and left detailed feedback above. I am happy to discuss or help debug. In the meantime, if any of you get a chance, I would appreciate a look at my open PR(s):

**My open PRs across OCA:**
- [server-tools#3554 [MIG] datetime_formatter: Migration to 18.0](https://github.com/OCA/server-tools/pull/3554)
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward. Thank you!

---
_Environment via [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer): Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

**Comment by nobuQuartile** [comment]
@marcos-mendez 
Those points have already been fixed.
I believe your review may have been based on an old code.

The error currently appearing has been fixed in this PR. https://github.com/OCA/manufacture/pull/1750
I believe it should be resolved after rebasing.

**Review by yostashiro — APPROVED** [review]
@nobuQuartile I've fixed some issues, expanded the scope to cover non-tracked products as well, and renamed the module. Please see my last commit.

Squash commits once you are okay with the content. Please update the PR title.

**Review by marcos-mendez — APPROVED** [review]
ok

**Comment by nobuQuartile** [comment]
@yostashiro 
It looks good. Thank you for your contribution.

> Squash commits once you are okay with the content. Please update the PR title.

Done

File: setup/mrp_unbuild_source_location/odoo/addons/mrp_unbuild_lot_location
Comment (yostashiro) [inline setup/mrp_unbuild_source_location/odoo/addons/mrp_unbuild_lot_location:1]: @nobuQuartile Not sure why this file is left here. Can you please remove this.

Comment (nobuQuartile) [reply setup/mrp_unbuild_source_location/odoo/addons/mrp_unbuild_lot_location:1]: Done. Thank you.
